### ISSUE TITLE
PEP 557 interop namespace

### ISSUE THREAD

**Issue by micimize** [python-attrs/attrs#686]
Extracted from https://github.com/python-attrs/attrs/issues/408#issuecomment-514375369

> On the other side of this, what about making a namespace that actually _is_ intentionally similar to PEP 557 `dataclasses` and `fields`, and just has a thin adapter layer between `field` and `attr.ib`, etc? How much incompatibility is there beyond the names? Isn't `attrs` mostly a superset?

It would be good for migrating from large codebases where `dataclasses` are already used, or for pythonistas who want codebases/apis that stick close to the PEP specifications where possible.

**Comment by wsanchez** [comment]
I think that if we did that, we should have an `attrs.dataclass` namespace, rather than using the primary namespace.

The reason for that is that if we start down that path and set and expectations that Attrs is PEP 557 (plus whatever extra awesome) then (a) we might get cornered into design decisions made elsewhere in order to maintain this expectation (b) we might have used a name that conflicts with some future dataclass addition and (c) compatibility may require contortions that would be better contained in a compatibility namespace.

**Comment by micimize** [comment]
agreed – I think it should actually be `attrs.dataclasses`, such that migrating from the standard lib is just adding the prefix:
`sed 's/from dataclasses import/from attrs.dataclasses import'`

**Comment by asford** [comment]
+1 to this idea, given what we've seen in https://github.com/python-attrs/attrs/issues/795, I think there are now more knock-on benefits than ever to adding a "dataclass compatible namespace". 

A few potential benefits would include:

* Direct support for `pyright` et al via the `dataclass_transform` spec, which is now [PEP 681](https://peps.python.org/pep-0681/)
* A smooth upgrade path for codebases that ~~see the light~~ wish to adopt attrs features.
* A smooth "reverse compatibility" path for codebases using attrs, but need to interact with dataclass-ish libraries.

The goal of the namespace would be to provide an attrs mode that is _superset_ of  `stdlib` dataclasses,
where an application can update to attrs by just changing an `import` statement.
This falls into two branches: differences in behavior for existing features and features in dataclasses not supported in attrs.
A non-exhaustive index includes:

Differences for existing features are blocking changes that prevent an application from directly converting from `dataclasses` to `attrs`. This will require optional changes to `attrs` behavior and are "must have" for this proposed namespace:

* `attrs` [private attributes handling](https://www.attrs.org/en/stable/init.html?highlight=private#private-attributes), which strips leading underscores from member names.
* `attrs` field ordering during inheritance, perhaps, doesn't match the dataclasses [ordering under inheritance](https://docs.python.org/3/library/dataclasses.html#inheritance)

Missing features are non-blocking changes that would require an application to reduce their use of `dataclasses` features to convert to `attrs`. These could be implemented as additional `attrs` features and are "nice to have" for this proposed namespace:
* `attrs` supports `__attrs_post_init__`, rather than `__post_init__`.
* `attrs` does not support [Init-only variables](https://docs.python.org/3/library/dataclasses.html#init-only-variables).
* `attrs` supports `factory`, but not `default_factory`, for default value factories. 
* `attrs` does not support `[KW_ONLY sentinel values](https://docs.python.org/3/library/dataclasses.html#dataclasses.KW_ONLY).

**Comment by asford** [comment]
On private attributes handling:

### Notes

[PEP681](https://peps.python.org/pep-0681/) introduces an optional "alias" parameter to field descriptors, which specifies an alternative `__init__` name for the member. This is not supported in dataclasses, but is [implemented in pydantic](https://pydantic-docs.helpmanual.io/usage/schema/#field-customization). 

A (simple?) solution may be to extend `attr.ib` with an optional `alias` parameter, which sets the `__init__` name for the field.
This would, as is the custom, be added to `attrs.Attribute` and would be available during for introspection in `attr.fields` and customization in the `field_transformer` hook.
If `alias` is unspecified it will be set to a `str`, indicating the attribute's argument name in the generated `__init__`.

For example in the current behavior, `alias = name.lstrip("_")` so that:
```python
>>> @attr.define
... class Foo:
...    bar: str
...    _bat: str
>>> attr.fields_dict(Foo)
... { "bar": Attribute(name="bar", alias="bar", ...), "_bat": Attribute(name="_bat", alias="bat", ...) }
```

`dataclass` compatibility could be introduced via an optional parameter in `attr.define` modifying the default alias behavior.
This may include a specific flag private attribute handling, or fold into an alternative flag controlling broad "dataclass compatibility". 

### Refs:

Related issues: 
#391
#244
#619
#171

This is currently implemented in `_make` via [`arg_name = a.name.lstrip("_")`](https://github.com/python-attrs/attrs/blob/main/src/attr/_make.py#L2250).
This is modeled in `evolve` via [`init_name = attr_name if attr_name[0] != "_" else attr_name[1:]`](https://github.com/python-attrs/attrs/blob/ae29da66b7c18928682679df1285deba95183689/src/attr/_funcs.py#L364).

The `lstrip`/`slice` inconsistency here is _almost certainly_ a bug for non-idiomatic cases like:

```python
@attr.define
class NonIdiomatic:
    ___triple_underscore: str
```

Exposing `alias` may be useful for libraries which interact with attrs-generated `__init__` methods. 
Typified by `cattrs`, which re-implement's private-attribute naming via [`ian = an if (is_dc or an[0] != "_") else an[1:]`](https://github.com/python-attrs/cattrs/blob/d946719320e27d015bbad03275fefe57b64b5869/src/cattrs/gen.py#L307).
Note, this implementation is also affected by `lstrip`/`slice` inconsistency.
This could be replaced by a lookup of the target Attribute alias.
